src/Interpreters/HashJoin.cpp

```diff
@@ -1122,94 +1166,155 @@ class AddedColumns
+                                dest_column->getName(), column_from_block->getName(),
+                                demangle(typeid(*dest_column).name()), demangle(typeid(*column_from_block).name()));
         }
-#endif
+    }
+
+    MutableColumns columns;
+    bool is_join_get;
+    std::vector<size_t> right_indexes;
+    std::vector<TypeAndName> type_name;
+    std::vector<ColumnNullable *> nullable_column_ptrs;
+    size_t lazy_defaults_count = 0;
+
+    /// for lazy
+    // The default row is represented by an empty RowRef, so that fixed-size blocks can be generated sequentially,
+    // default_count cannot represent the position of the row
+    LazyOutput lazy_output;
+    bool has_columns_to_add;
 
-        if (is_join_get)
+    /// for ASOF
+    const IColumn * left_asof_key = nullptr;
+
+
+    void addColumn(const ColumnWithTypeAndName & src_column, const std::string & qualified_name)
+    {
+        columns.push_back(src_column.column->cloneEmpty());
+        columns.back()->reserve(src_column.column->size());
+        type_name.emplace_back(src_column.type, src_column.name, qualified_name);
+    }
+};
+template<> void AddedColumns<false>::buildOutput()
+{
+}
+
+template<>
+void AddedColumns<true>::buildOutput()
+{
+    for (size_t i = 0; i < this->size(); ++i)
+    {
+        auto& col = columns[i];
+        size_t default_count = 0;
+        auto apply_default = [&]()
         {
-            size_t right_indexes_size = right_indexes.size();
-            for (size_t j = 0; j < right_indexes_size; ++j)
+            if (default_count > 0)
             {
-                const auto & column_from_block = block.getByPosition(right_indexes[j]);
-                if (auto * nullable_col = nullable_column_ptrs[j])
-                    nullable_col->insertFromNotNullable(*column_from_block.column, row_num);
-                else
-                    columns[j]->insertFrom(*column_from_block.column, row_num);
+                JoinCommon::addDefaultValues(*col, type_name[i].type, default_count);
+                default_count = 0;
             }
-        }
-        else
+        };
+
+        for (size_t j = 0; j < lazy_output.blocks.size(); ++j)
         {
-            size_t right_indexes_size = right_indexes.size();
-            for (size_t j = 0; j < right_indexes_size; ++j)
+            if (!lazy_output.blocks[j])
             {
-                const auto & column_from_block = block.getByPosition(right_indexes[j]);
-                columns[j]->insertFrom(*column_from_block.column, row_num);
+                default_count ++;
+                continue;
+            }
+            apply_default();
+            const auto & column_from_block = reinterpret_cast<const Block *>(lazy_output.blocks[j])->getByPosition(right_indexes[i]);
+            /// If it's joinGetOrNull, we need to wrap not-nullable columns in StorageJoin.
+            if (is_join_get)
```

Comment:
I assume `is_join_get = false` in all cases, because we call `joinBlockImpl` with `Any` for dictGet. But I'm not 100% sure, also if something changed we will get a wrong behavior, so we can keep this `if`.

Reply:
I don’t particularly understand the logic here. In order not to cause unnecessary bugs, I have retained this logic.